src/WebJobs.Extensions.DurableTask/Listener/TaskOrchestrationShim.cs

```diff
@@ -117,7 +124,60 @@ public void SetFunctionInvocationCallback(Func<Task> callback)
 
             if (returnValue != null)
             {
-                this.context.SetOutput(returnValue);
+                if (orchestratorInfo.IsOutOfProc && returnValue is JObject)
+                {
+                    var jObj = returnValue as JObject;
+                    if (jObj["actions"] != null)
+                    {
+                        var commandSets = JsonConvert.DeserializeObject<AsyncCommand[][]>(jObj["actions"].ToString());
+
+                        foreach (AsyncCommand[] commandSet in commandSets)
+                        {
+                            var tasks = new List<Task>();
+                            foreach (AsyncCommand command in commandSet)
+                            {
+                                switch (command.ActionType)
+                                {
+                                    case AsyncActionType.CallActivity:
+                                        tasks.Add(this.context.CallActivityAsync(command.FunctionName, command.Input));
+                                        break;
+                                    case AsyncActionType.CreateTimer:
+                                        using (var cts = new CancellationTokenSource())
+                                        {
+                                            tasks.Add(this.context.CreateTimer(command.FireAt, cts.Token));
+                                            if (command.IsCanceled)
+                                            {
+                                                cts.Cancel();
+                                            }
+                                        }
+
+                                        break;
+                                    case AsyncActionType.CallActivityWithRetry: break;
+                                    case AsyncActionType.CallSubOrchestrator: break;
+                                    case AsyncActionType.CallSubOrchestratorWithRetry: break;
+                                    case AsyncActionType.ContinueAsNew: break;
+                                    case AsyncActionType.WaitForExternalEvent:
+                                        tasks.Add(this.context.WaitForExternalEvent<object>(command.ExternalEventName));
+                                        break;
+                                    default:
+                                        break;
+                                }
+                            }
+
+                            await Task.WhenAny(tasks);
+                        }
+                    }
+
+                    var isDone = bool.Parse(jObj["isDone"].ToString());
+                    if (isDone)
+                    {
+                        this.Context.SetOutput(jObj["output"]);
+                    }
+                }
+                else
+                {
+                    this.context.SetOutput(returnValue);
+                }
```

Comment:
This is a big block of new logic specific to out-of-proc . Would it make sense to factor it into a separate method for improved readability?

Reply:
It did get rather large. I'll move it into a method.